src/apps/webhooks.clj

```diff
@@ -0,0 +1,61 @@
+(ns apps.webhooks
+  (:use [korma.core :exclude [update]]
+        [korma.db :only [transaction]]
+        [apps.persistence.users :only [get-user-id]])
+  (:require [clojure.tools.logging :as log]
+            [clojure-commons.exception-util :as cxu]
+            [apps.util.config :as config]
+            [korma.core :as sql]))
+
+(defn- get-webhook-topics [webhook-id]
+  (map :topic (select [:webhooks_topic :wt]
+                      (join [:webhooks_subscription :ws] {:wt.id :ws.topic_id})
+                      (fields [:wt.topic])
+                      (where {:ws.webhook_id webhook-id}))))
+
+(defn- format-webhook [{:keys [id] :as webhook}]
+  (assoc webhook :topics (get-webhook-topics id)))
+
+(defn get-webhooks [user]
+  (transaction
+    (let [webhooks (select [:webhooks :w]
+                           (join [:users :u] {:w.user_id :u.id})
+                           (join [:webhooks_type :wt] {:w.type_id :wt.id})
+                           (fields [:w.id :id] [:wt.type :type] [:w.url :url])
+                           (where {(sqlfn regexp_replace :u.username "@.*" "")
+                                   user}))]
+      (map format-webhook webhooks))))
+
+(defn list-webhooks [user]
+  {:webhooks (get-webhooks user)})
+
+(defn get-type-id [type]
+  (if-let [type_id (:id (first (select :webhooks_type (fields :id) (where {:type type}))))]
+    type_id
+    (cxu/bad-request (str "Unable to find webhook type: " type))))
+
+(defn- add-topic-subscription [webhook_id topic_name]
+  (let [topic_id (:id (first (select :webhooks_topic (fields :id) (where {:topic topic_name}))))]
+    (when topic_id
+      (insert :webhooks_subscription (values {:webhook_id (log/spy webhook_id)
+                                              :topic_id   topic_id})))))
+(defn add-webhook [user url type topics]
+  (let [user_id (get-user-id (str user "@" (config/uid-domain)))
```

Comment:
Since `add-webhook` is only being called by `add-webhooks`, and `add-webhooks` is already looking up the `user_id`, then it could be passed directly to this function instead of the username. That way `get-user-id` only has to be called once, and not for every webhook to be added.

Reply:
👍